src/Functions/array/arrayRandomSample.cpp

```diff
@@ -0,0 +1,143 @@
+#include <random>
+#include <Columns/ColumnArray.h>
+#include <DataTypes/DataTypeArray.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/IFunction.h>
+#include <Poco/Logger.h>
+#include "Columns/ColumnsNumber.h"
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_COLUMN;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+/// arrayRandomSample(arr, k) - Returns k random elements from the input array
+class FunctionArrayRandomSample : public IFunction
+{
+public:
+    static constexpr auto name = "arrayRandomSample";
+
+    static FunctionPtr create(ContextPtr) { return std::make_shared<FunctionArrayRandomSample>(); }
+
+    String getName() const override { return name; }
+
+    size_t getNumberOfArguments() const override { return 2; }
+    bool useDefaultImplementationForConstants() const override { return true; }
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        const DataTypePtr & first_arg = arguments[0];
+        if (!isArray(first_arg))
+            throw Exception(
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                "Illegal type {} of argument of function {}, expected Array",
+                arguments[0]->getName(),
+                getName());
+
+        const DataTypeArray * array_type = checkAndGetDataType<const DataTypeArray>(first_arg.get());
+
+        if (!isUnsignedInteger(arguments[1]))
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Second argument must be a unsigned integer");
+
+        // Return an array with the same nested type as the input array
+        return std::make_shared<DataTypeArray>(array_type->getNestedType());
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t) const override
+    {
+        ColumnPtr column_array_ptr = arguments[0].column;
+        const auto * column_array = checkAndGetColumn<ColumnArray>(column_array_ptr.get());
+
+        if (!column_array)
+        {
+            const auto * column_const_array = checkAndGetColumnConst<ColumnArray>(column_array_ptr.get());
+            if (!column_const_array)
+                throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Expected Array column, found {}", column_array_ptr->getName());
+
+            column_array_ptr = column_const_array->convertToFullColumn();
+            column_array = assert_cast<const ColumnArray *>(column_array_ptr.get());
+        }
```

Comment:
@rschu1ze, these codes fix the materialize function option, but now I see two rows in the result for some reason.
I am checking it. 

See this one for example:
```
home :) select arrayRandomSample(['a', 'b', 'c'], materialize(1));

SELECT arrayRandomSample(['a', 'b', 'c'], materialize(1))

Query id: 4b497658-8b39-4bcf-b4af-64a2f6137c6e

┌─arrayRandomSample(['a', 'b', 'c'], materialize(1))─┐
│ []                                                 │
└────────────────────────────────────────────────────┘
┌─arrayRandomSample(['a', 'b', 'c'], materialize(1))─┐
│ ['b']                                              │
└────────────────────────────────────────────────────┘

2 rows in set. Elapsed: 0.004 sec. 
```

Reply:
Weird. `select arrayRandomSample(['a', 'b', 'c'], materialize(1));` should throw an exception.

Reply:
I suspect these lines:
```
column_array_ptr = column_const_array->convertToFullColumn();
column_array = assert_cast<const ColumnArray *>(column_array_ptr.get());
```

Reply:
In the end, I do not block samples to by only const, block it? 
I did the same code as array param:
```
ColumnPtr col_num_ptr = arguments[1].column;
        const IColumn * col_num = col_num_ptr.get();

        if (!col_num)
        {
            const auto * column_const_num = checkAndGetColumnConst<ColumnConst>(col_num_ptr.get());
            if (!column_const_num || column_const_num->empty())
                throw Exception(
                    ErrorCodes::ILLEGAL_COLUMN, "The second argument column is empty or null, type = {}", arguments[1].type->getName());
            col_num_ptr = column_const_num->convertToFullColumn();
            col_num = col_num_ptr.get();
        }
```

Reply:
I am getting some errors; let's solve this problem with a separate PR. 
because my time is a little limited for the following week.